[Runtime] Handle Error-conforming-to-NSObject casting fully.

DougGregor opened the pull request:
Rather than attempting Error bridging early when trying to dynamically
cast to NSError or NSObject, treat it as the *last* thing we do when
all else fails. Push most of this code over into Objective-C-specific
handling rather than #ifdef'd into the main casting logic to make that
slightly more clear.

One oddity of Error/NSError bridging is that a class that conforms to
Error can be dynamically cast to NSObject via Error bridging. This has
always been known to the static compiler, but the runtime itself was
not always handling such a cast uniformly. Do so now,
uniformly. However, this forced us to weaken an assertion, because
casting a class type to NSError or NSObject can produce an object with
a different identity.

Fixes rdar://problem/57393991.

DougGregor committed:
[Test] Go through the runtime for the "as? NSObject" check.

This prevents runtime differences on as? from affecting this test.

DougGregor committed:
[Runtime] Handle Error-conforming-to-NSObject casting fully.

I missed a case where an Error-conforming class is dynamically casted
to NSObject (via NSError). Fix it.

DougGregor commented:
@swift-ci please test

mikeash commented:
Looks OK to me! I also added @tbkka as a reviewer. At this point he probably knows this code better than I do.

You should also revert the disabled test from https://github.com/apple/swift/pull/28406.

DougGregor committed:
Revert "Disable a test to unblock CI"

This reverts commit 29a6d1066e765cdb3a5e1d7bdfb22310b2885664.

DougGregor commented:
Yikes, I thought #28406 had already been reverted with the rest. And it's still unfixed, *thank you*.

mikeash commented:
D'oh. :(

DougGregor committed:
[Runtime] Handle Error bridging as a last chance to cast to NSError/NSObject.

Rather than attempting Error bridging early when trying to dynamically
cast to NSError or NSObject, treat it as the *last* thing we do when
all else fails. Push most of this code over into Objective-C-specific
handling rather than #ifdef'd into the main casting logic to make that
slightly more clear.

One oddity of Error/NSError bridging is that a class that conforms to
Error can be dynamically cast to NSObject via Error bridging. This has
always been known to the static compiler, but the runtime itself was
not always handling such a cast uniformly. Do so now,
uniformly. However, this forced us to weaken an assertion, because
casting a class type to NSError or NSObject can produce an object with
a different identity.

Fixes rdar://problem/57393991.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
@swift-ci please test platform macOS

DougGregor commented:
@swift-ci please test platform macOS

DougGregor commented:
@swift-ci please test Linux

DougGregor commented:
I'm going to merge because other PRs will be built on top of this, but we'll back it out of there are any concerns.

mikeash commented:
Looks good!
